File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/booking/BookingService.java
package at.ac.tuwien.sepm.groupphase.backend.service.booking;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.booking.*;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.user.BookingOverviewDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Booking;
import at.ac.tuwien.sepm.groupphase.backend.repository.BookedNonSeat;
import at.ac.tuwien.sepm.groupphase.backend.repository.BookedSeat;
import java.util.List;

public interface BookingService {

  /**
   * Creates a booking and invoice entry in the system and either a booking-seat or -non seat entry.
   *
   * @param reservationDto stores booking information
   * @return booking entity storing relevant information about the reservation
   */
  Booking reserve(ReservationDto reservationDto, Long eventShowingId);

  /**
   * Creates a booking and invoice entry in the system and either a booking-seat or -non seat entry.
   *
   * @param purchaseDto stores booking information as well as payment information
   * @return booking entity storing relevant information about the purchase
   */
  Booking purchase(PurchaseDto purchaseDto, Long eventShowingId);

  /**
   * Gets all information necessary to turn a reservation into a purchase.
   *
   * @param bookingId matching the reservation
   * @return information about the reservation
   */
  ReservationInfoDto getReservationInfo(Long bookingId);

  /**
   * Creates an invoice for type purchase for an already existing booking.
   *
   * @param bookingId referencing a booking being turned into a purchase
   * @param reservationPurchaseDto consists of booked seats, non seats and payment information
   * @return invoice entity storing relevant information about the purchase
   */
  Booking reservationToPurchase(Long bookingId, ReservationPurchaseDto reservationPurchaseDto);

  /**
   * Returns all seats, that are already booked for this showing.
   *
   * @param showingId id of the showing
   * @return List of all seats that were booked
   */
  List<BookedSeat> findBookedSeats(Long showingId);

  /**
   * Returns all non-seats, that are already booked for this showing.
   *
   * @param showingId id of the showing
   * @return List of all non-seats that were booked
   */
  List<BookedNonSeat> findBookedNonSeats(Long showingId);

  /**
   * Returns all bookings with bookedBy.
   *
   * @param bookedBy used for matching bookedBy in database
   * @return List of BookingDtos
   */
  List<BookingOverviewDto> findAllBookingsByBookedBy(Long bookedBy);

  /**
   * Returns details about the showing with the given id.
   *
   * @param id the showing id
   * @return Details to the showing
   */
  BookingShowingDetailsDto findShowingDetailsById(Long id);

  /**
   * Cancels one booking with the given bookingId and returns a BookingOverviewDto with the changed
   * is_canceled.
   *
   * @param bookingId the booking id
   * @return BookingOverviewDto
   */
  BookingOverviewDto cancelBooking(Long bookingId);
}

File: frontend/src/app/services/booking/booking.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Globals} from '../../global/globals';
import {ReservationDto} from '../../dtos/booking/reservation.dto';
import {Observable} from 'rxjs';
import {BookingDto} from '../../dtos/booking/booking.dto';
import {PurchaseDto} from '../../dtos/booking/purchase.dto';
import {BookingOverviewDto} from 'src/app/dtos/booking/booking-overview.dto';
import {ShowingDetails} from 'src/app/dtos/booking/showing-details.dto';
import {ReservationInfoDto} from '../../dtos/booking/reservation-info.dto';
import {ReservationPurchaseDto} from '../../dtos/booking/reservation-purchase.dto';

@Injectable({
  providedIn: 'root'
})
export class BookingService {

  private bookingBaseUri: string = this.globals.backendUri + '/showing';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Creates booking
   *
   * @param reservationDto all information about reservation being made
   * @param eventShowingId showing being reserved
   */
  reserve(reservationDto: ReservationDto, eventShowingId: number): Observable<BookingDto> {
    const uri = this.bookingBaseUri + '/' + eventShowingId + '/reservation';
    return this.httpClient.post<BookingDto>(uri, reservationDto);
  }

  /**
   * Creates booking and invoice
   *
   * @param purchaseDto all information about purchase being made
   * @param eventShowingId showing being purchased
   */
  purchase(purchaseDto: PurchaseDto, eventShowingId: number): Observable<BookingDto> {
    const uri = this.bookingBaseUri + '/' + eventShowingId + '/purchase';
    return this.httpClient.post<BookingDto>(uri, purchaseDto);
  }

  /**
   * Cancels a booking with bookingId.
   *
   * @param bookingId id of the booking
   */
  cancel(bookingId: number): Observable<BookingOverviewDto> {
    return this.httpClient.patch<BookingOverviewDto>(this.globals.backendUri + '/booking/' + bookingId + '/cancellation', bookingId);
  }

  /**
   * Gets all relevant information about a prior made reservation
   *
   * @param bookingId of reservation
   */
  getReservationInfo(bookingId: number): Observable<ReservationInfoDto> {
    const uri = this.globals.backendUri + '/booking/' + bookingId + '/reservation-info';
    return this.httpClient.get<ReservationInfoDto>(uri);
  }

  /**
   * Creates an invoice for an already existing booking (turns reservation into purchase)
   *
   * @param bookingId of already existing reservation
   * @param reservationPurchaseDto consisting of purchase information
   */
  reservationToPurchase(bookingId: number, reservationPurchaseDto: ReservationPurchaseDto): Observable<BookingDto> {
    const uri = this.globals.backendUri + '/booking/' + bookingId + '/reservation-to-purchase';
    return this.httpClient.post<BookingDto>(uri, reservationPurchaseDto);
  }

  /**
   * Gets bookings of user with id.
   *
   * @param id id of the user
   */
  public getBookingsByUser(id: number): Observable<BookingOverviewDto[]> {
    return this.httpClient.get<BookingOverviewDto[]>(this.globals.backendUri + '/user/' + id + '/bookings');
  }

  /**
   * Gets details of the showing with this id.
   *
   * @param showingId the id of the showing
   * @returns details about the showing
   */
  public getShowingDetails(showingId: number): Observable<ShowingDetails> {
    const uri = this.globals.backendUri + '/showing/' + showingId + '/details';
    return this.httpClient.get<ShowingDetails>(uri);
  }
}
